crates/noirc_frontend/src/hir_def/types.rs

```diff
@@ -387,55 +392,82 @@ impl Type {
         }
     }
 
-    /// Returns true on success
+    pub fn set_const(&mut self, new_const: IsConst) {
+        match self {
+            Type::FieldElement(is_const) | Type::Integer(is_const, _, _) => {
+                *is_const = new_const;
+            }
+            Type::PolymorphicInteger(is_const, binding) => {
+                if let TypeBinding::Bound(binding) = &mut *binding.borrow_mut() {
+                    return binding.set_const(new_const);
+                }
+                *is_const = new_const;
+            }
+            _ => (),
+        }
+    }
+
+    /// Try to bind a PolymorphicInt variable to self, succeeding if self is an integer, field,
+    /// other PolymorphicInt type, or type variable. If use_subtype is true, the IsConst fields
+    /// of each will be checked via subtyping rather than unification.
     pub fn try_bind_to_polymorphic_int(
         &self,
         var: &TypeVariable,
         var_const: &IsConst,
+        use_subtype: bool,
         span: Span,
     ) -> Result<(), SpanKind> {
         let target_id = match &*var.borrow() {
             TypeBinding::Bound(_) => unreachable!(),
             TypeBinding::Unbound(id) => *id,
         };
 
-        match self {
-            Type::FieldElement(is_const, ..) | Type::Integer(is_const, ..) => {
-                let mut clone = self.clone();
-                clone.set_const_span(span);
-                *var.borrow_mut() = TypeBinding::Bound(clone);
-                is_const.unify(var_const, span)
+        let bind = |int_const: &IsConst| {
+            let mut clone = self.clone();
```

Comment:
Several copies are done here, is there a reason why we cannot modify the data in-place?

Reply:
Yes, when we bind a type variable to some new type we have to do `*var.borrow_mut() = TypeBinding::Bound(new_type);` to mutate it since there are potentially N different references to any given type variable and they must all point to the new binding. Mutating something locally would not change these other copies, and would still require we cloned the data to mutate to.